pkg/frontend/v2/frontend_test.go

```diff
@@ -332,8 +339,82 @@ func TestFrontendFailedCancellation(t *testing.T) {
 	require.Nil(t, resp)
 
 	ms.checkWithLock(func() {
-		require.Equal(t, 1, len(ms.msgs))
+		require.Len(t, ms.msgs, 1)
+	})
+}
+
+// Test that the front-end shuts down gracefully, i.e. waits for in-flight queries to finish.
+func TestFrontend_GracefulShutdown(t *testing.T) {
+	const (
+		userID = "test"
+		body   = "all good"
+	)
+	ctx := user.InjectOrgID(context.Background(), userID)
+
+	setupDone := make(chan uint64)
+	t.Cleanup(func() {
+		close(setupDone)
 	})
+	f, _ := setupFrontend(t, nil, func(f *Frontend, msg *schedulerpb.FrontendToScheduler) *schedulerpb.SchedulerToFrontend {
+		go func() {
+			// Allow for the service to register a successful enqueuing
+			time.Sleep(100 * time.Millisecond)
+			setupDone <- msg.QueryID
+		}()
```

Comment:
I don't understand this:
- Why run in a dedicated goroutine?
- Why the sleep?

Reply:
@pracucci I've followed the existing logic in [TestFrontendBasicWorkflow](https://github.com/grafana/mimir/blob/main/pkg/frontend/v2/frontend_test.go#L123) (including the sleep), just that I don't send the response as part of this goroutine.

My understanding is to sleep in order to let the mock scheduler register the `schedulerpb.OK` status, and then let the main test goroutine know that setup is over (via `setupDone`).